clients/cli/cmd/internal/pull.go

```diff
@@ -155,47 +157,146 @@ func (target *Target) DownloadAndWriteToFile(client *phrase.APIClient, localeFil
+	}
+}
+
+func (target *Target) downloadAsynchronously(client *phrase.APIClient, localeFile *LocaleFile, downloadOpts phrase.LocaleDownloadOpts) error {
+	localeDownloadCreateParams := asyncDownloadParams(downloadOpts)
+
+	localVarOptionals := phrase.LocaleDownloadCreateOpts{}
+	debugFprintln("Initiating async download...")
+	asyncDownload, _, err := client.LocaleDownloadsApi.LocaleDownloadCreate(Auth, target.ProjectID, localeFile.ID, localeDownloadCreateParams, &localVarOptionals)
+	if err != nil {
+		return err
+	}
+
+	for asyncDownload.Status == "processing" {
+		debugFprintln("Waiting for the files to be exported...")
+		time.Sleep(asyncWaitTime)
+		debugFprintln("Checking if the download is ready...")
+		localVarOptionals := phrase.LocaleDownloadShowOpts{}
+		asyncDownload, _, err = client.LocaleDownloadsApi.LocaleDownloadShow(Auth, target.ProjectID, localeFile.ID, asyncDownload.Id, &localVarOptionals)
+		if err != nil {
+			return err
+		}
+	}
+	if asyncDownload.Status == "failed" {
+		return fmt.Errorf("download failed: %s", asyncDownload.Error)
+	}
+	return downloadExportedLocale(asyncDownload.Result.Url, localeFile.Path)
+}
+
+func (target *Target) downloadSynchronously(client *phrase.APIClient, localeFile *LocaleFile, downloadOpts phrase.LocaleDownloadOpts) error {
+	file, response, err := client.LocalesApi.LocaleDownload(Auth, target.ProjectID, localeFile.ID, &downloadOpts)
 	if err != nil {
 		if response.Rate.Remaining == 0 {
 			waitForRateLimit(response.Rate)
-			file, _, err = client.LocalesApi.LocaleDownload(Auth, target.ProjectID, localeFile.ID, &localVarOptionals)
+			file, _, err = client.LocalesApi.LocaleDownload(Auth, target.ProjectID, localeFile.ID, &downloadOpts)
 			if err != nil {
 				return err
 			}
 		} else {
 			return err
 		}
 	}
+	return copyToDestination(file, localeFile.Path)
+}
 
+func copyToDestination(file *os.File, path string) error {
 	var data []byte
-	if file != nil {
-		data, err = ioutil.ReadAll(file)
-		if err != nil {
-			return err
-		}
-		file.Close()
-		os.Remove(file.Name())
+	data, err := io.ReadAll(file)
+	if err != nil {
+		return err
 	}
+	file.Close()
+	os.Remove(file.Name())
 
-	err = ioutil.WriteFile(localeFile.Path, data, 0644)
+	err = os.WriteFile(path, data, 0644)
 	return err
 }
 
+func downloadExportedLocale(url string, localName string) error {
+	debugFprintln("Downloading file from ", url)
+	file, err := os.Create(localName)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	request, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+	request.Header.Set("Authorization", "Bearer "+Config.Credentials.Token)
+	response, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	io.Copy(file, response.Body)
+	return nil
+}
+
+func asyncDownloadParams(localVarOptionals phrase.LocaleDownloadOpts) phrase.LocaleDownloadCreateParameters {
```

Comment:
I think we have to somehow automatically convert `phrase.LocaleDownloadOpts` to `phrase.LocaleDownloadCreateParameters`. Otherwise we will always play catch-up when a new download option is introduced and will have to add it here

Reply:
with a great help of Copilot :pilot: :sweat_smile: , I managed to rewrite it to use reflection. I tried some basic cases and it seems to work.